code/modules/food_and_drinks/kitchen_machinery/coffeemaker.dm

```diff
@@ -0,0 +1,693 @@
+		return
+	if(length(options) == 1)
+		choice = options[1]
+	else
+		choice = show_radial_menu(user, src, options, require_near = !issilicon(user))
+
+	// post choice verification
+	if(brewing || panel_open || !anchored || !user.canUseTopic(src, !issilicon(user)))
+		return
+
+	switch(choice)
+		if("Brew")
+			brew(user)
+		if("Eject Pot")
+			eject_pot(user)
+		if("Eject Cartridge")
+			eject_cartridge(user)
+		if("Examine")
+			examine(user)
+		if("Take Cup")
+			take_cup(user)
+		if("Take Sugar")
+			take_sugar(user)
+		if("Take Sweetener")
+			take_sweetener(user)
+		if("Take Creamer")
+			take_creamer(user)
+
+/obj/machinery/coffeemaker/proc/eject_pot(mob/user)
+	if(coffeepot)
+		replace_pot(user)
+
+/obj/machinery/coffeemaker/proc/eject_cartridge(mob/user)
+	if(cartridge)
+		replace_cartridge(user)
+
+/obj/machinery/coffeemaker/proc/take_cup(mob/user)
+	if(!coffee_cups) //shouldn't happen, but we all know how stuff manages to break
+		balloon_alert(user, "no cups left!")
+		return
+	var/obj/item/reagent_containers/glass/coffee_cup/new_cup = new(get_turf(src))
+	user.put_in_hands(new_cup)
+	coffee_cups--
+	update_appearance(UPDATE_OVERLAYS)
+
+/obj/machinery/coffeemaker/proc/take_sugar(mob/user)
+	if(!sugar_packs)
+		balloon_alert(user, "no sugar left!")
+		return
+	var/obj/item/reagent_containers/food/condiment/pack/sugar/new_pack = new(get_turf(src))
+	user.put_in_hands(new_pack)
+	sugar_packs--
+	update_appearance(UPDATE_OVERLAYS)
+
+/obj/machinery/coffeemaker/proc/take_sweetener(mob/user)
+	if(!sweetener_packs)
+		balloon_alert(user, "no sweetener left!")
+		return
+	var/obj/item/reagent_containers/food/condiment/pack/astrotame/new_pack = new(get_turf(src))
+	user.put_in_hands(new_pack)
+	sweetener_packs--
+	update_appearance(UPDATE_OVERLAYS)
+
+/obj/machinery/coffeemaker/proc/take_creamer(mob/user)
+	if(!creamer_packs)
+		balloon_alert(user, "no creamer left!")
+		return
+	var/obj/item/reagent_containers/food/condiment/pack/creamer/new_pack = new(drop_location())
+	user.put_in_hands(new_pack)
+	creamer_packs--
+	update_appearance(UPDATE_OVERLAYS)
+
+///Updates the smoke state to something else, setting particles if relevant
+/obj/machinery/coffeemaker/proc/toggle_steam()
+	QDEL_NULL(particles)
+	if(brewing)
+		particles.position = list(-6, 0, 0)
+
+/obj/machinery/coffeemaker/proc/operate_for(time, silent = FALSE)
+	brewing = TRUE
+	if(!silent)
+		playsound(src, 'sound/machines/coffeemaker_brew.ogg', 20, vary = TRUE)
+	toggle_steam()
+	use_power(active_power_usage * time * 0.1) // .1 needed here to convert time (in deciseconds) to seconds such that watts * seconds = joules
+	addtimer(CALLBACK(src, PROC_REF(stop_operating)), time / speed)
+
+/obj/machinery/coffeemaker/proc/stop_operating()
+	brewing = FALSE
+	toggle_steam()
+
+/obj/machinery/coffeemaker/proc/brew()
+	power_change()
+	if(!try_brew())
+		return
+	operate_for(brew_time)
+	coffeepot.reagents.add_reagent_list(cartridge.drink_type)
+	cartridge.charges--
```

Comment:
this doesn't take a `mob/user` arg even though it's passed, and doesn't continue to pass it to `try_brew()`, causing it to have to use `usr`. it's best to just pass the var along

Reply:
Ill be real honest, not sure how to do this one